geode-core/src/main/java/org/apache/geode/internal/net/BufferPool.java

```diff
@@ -67,39 +68,18 @@ public ByteBuffer acquireReceiveBuffer(int size) {
   private ByteBuffer acquireBuffer(int size, boolean send) {
     ByteBuffer result;
     if (useDirectBuffers) {
-      IdentityHashMap<BBSoftReference, BBSoftReference> alreadySeen = null; // keys are used like a
-                                                                            // set
-      BBSoftReference ref = bufferQueue.poll();
-      while (ref != null) {
-        ByteBuffer bb = ref.getBB();
-        if (bb == null) {
-          // it was garbage collected
-          int refSize = ref.consumeSize();
-          if (refSize > 0) {
-            if (ref.getSend()) { // fix bug 46773
-              stats.incSenderBufferSize(-refSize, true);
-            } else {
-              stats.incReceiverBufferSize(-refSize, true);
-            }
-          }
-        } else if (bb.capacity() >= size) {
-          bb.rewind();
-          bb.limit(size);
-          return bb;
-        } else {
-          // wasn't big enough so put it back in the queue
-          Assert.assertTrue(bufferQueue.offer(ref));
-          if (alreadySeen == null) {
-            alreadySeen = new IdentityHashMap<>();
-          }
-          if (alreadySeen.put(ref, ref) != null) {
-            // if it returns non-null then we have already seen this item
-            // so we have worked all the way through the queue once.
-            // So it is time to give up and allocate a new buffer.
-            break;
-          }
+      for(ByteBufferSoftReference ref : bufferQueue) {
```

Comment:
GEODE-6636 wanted this iteration to be done in a way that would only remove items from the queue if they did not need to be added back. But this code still adds back on line 81 (offerFirst). The idea of the EqualsWithPredicate is to do all the tests in the predicate and only remove an item that match the predicate. So the predicate could do the size check. The garbage collection one is trickier. We basically would want to skip it. We could keep count of how many are gc'd and once that hits some threshold do a scan that just removes the null items from the queue.

Reply:
I can see how this would create more garbage. Our concern was that if we left the buffer in the queue, then this method would not be thread safe without additional synchronization. We were reluctant to make the method synchronized because we didn't want to increase the amount of time spent spinning on waiting to get into the method. But we should look into splitting the difference, only taking things off the queue when we need to, and synchronizing around the smallest block of code possible.

Reply:
If you end up sticking with removing it early and adding it back, then consider some other way of iterating. Currently you have the iterator (in the "for" statement) and then on each call of remove(Object) it does another iteration (starting at head) until it finds your item.
It seems like you could just use an Iterator on it explicitly (instead of the implicit on with "for") and then use the remove method on the iterator.